### ISSUE TITLE
[Project]: Add Multimodal Hate Speech Detection using Text + Image Fusion (DistilBERT + ResNet-50)

### ISSUE THREAD

**Issue by divyanshim27** [abhisheks008/DL-Simplified#1198]
## Summary

I would like to propose and implement a new project: **Multimodal Hate Speech Detection using Text and Image Fusion.** While DL-Simplified has several NLP-based hate speech or sentiment projects, none of them tackle the multimodal problem — where the hateful signal appears only when text and image are considered together (e.g., a benign image with hateful caption, or a meme). This is an advanced, real-world problem with direct societal impact and introduces contributors to multimodal deep learning, an important and growing subfield.

## Why This Project

- The Facebook Hateful Memes Challenge (2020) brought this problem to mainstream attention — it remains an active research area.
- DL-Simplified has no existing multimodal (text + vision) fusion project.
- This introduces contributors to cross-modal attention, late fusion, and early fusion architectures — concepts not covered in any current project.
- The dataset (Facebook Hateful Memes) is publicly available with a Kaggle mirror.

## Proposed Model Architecture

**Late Fusion approach (three models compared in the notebook):**

```python
import torch
import torch.nn as nn
from transformers import DistilBertModel
from torchvision.models import resnet50, ResNet50_Weights

class MultimodalHateSpeechDetector(nn.Module):
    def __init__(self):
        super().__init__()
        # Text encoder
        self.text_encoder = DistilBertModel.from_pretrained('distilbert-base-uncased')
        self.text_proj = nn.Linear(768, 256)

        # Image encoder (frozen backbone + fine-tuned head)
        resnet = resnet50(weights=ResNet50_Weights.DEFAULT)
        self.image_encoder = nn.Sequential(*list(resnet.children())[:-1])
        self.image_proj = nn.Linear(2048, 256)

        # Fusion classifier
        self.fusion = nn.Sequential(
            nn.Linear(512, 128),
            nn.ReLU(),
            nn.Dropout(0.4),
            nn.Linear(128, 1),
            nn.Sigmoid()
        )

    def forward(self, input_ids, attention_mask, pixel_values):
        text_feat = self.text_encoder(input_ids, attention_mask).last_hidden_state[:, 0]
        text_feat = self.text_proj(text_feat)

        img_feat = self.image_encoder(pixel_values).squeeze(-1).squeeze(-1)
        img_feat = self.image_proj(img_feat)

        fused = torch.cat([text_feat, img_feat], dim=-1)
        return self.fusion(fused).squeeze(-1)
```

**Three architectures compared:**
1. Text-only: DistilBERT classifier (baseline)
2. Image-only: ResNet-50 classifier (baseline)
3. Multimodal: Late fusion of both (proposed)

**Expected results:** Multimodal AUROC ~0.72–0.76 on the Hateful Memes eval set, improving over either unimodal baseline.

**Notebook structure (following repo template):**

Multimodal Hate Speech Detection/
├── Dataset/
│ └── README.md # Hateful Memes download link, license
├── Images/
│ ├── model_architecture.png
│ ├── confusion_matrix.png
│ ├── roc_auc_comparison.png
│ └── sample_predictions.png
├── Model/
│ ├── Multimodal_Hate_Speech_Detection.ipynb
│ └── README.md
└── requirements.txt

**Libraries:** `transformers`, `torch`, `torchvision`, `Pillow`, `scikit-learn`, `matplotlib`, `seaborn`, `pandas`

## Additional Notes

- The notebook will include a clear explanation of the multimodal problem that beginners can follow.
- An ethical disclaimer section will address responsible use and model limitations.
- A Gradio demo will allow users to test with a custom image + text pair.

Could you please assign this issue to me?

**Labels:** `enhancement`, `new-project`, `nlp`, `computer-vision`, `multimodal`, `advanced`, `help wanted`, `GSSoC 2026`
